File: dotnet/src/SodiumFRP/Sodium.Tests.Internal/CellTests.cs
using System;
using System.Collections.Generic;
using NUnit.Framework;

namespace Sodium.Tests.Internal
{
    [TestFixture]
    public class CellTests
    {
        [Test]
        public void TestTransaction()
        {
            bool calledBack = false;
            Transaction.Apply(
                (trans, _) =>
                {
                    trans.Prioritized(Node<Unit>.Null, trans2 => calledBack = true);
                    return Unit.Value;
                },
                false);
            Assert.IsTrue(calledBack);
        }

        [Test]
        public void TestRegen()
        {
            List<int> @out = new List<int>();

            Transaction.Apply(
                (trans, _) =>
                {
                    void SetNeedsRegeneratingAndPrioritized(Action action)
                    {
                        trans.SetNeedsRegenerating();
                        trans.Prioritized(new Node<Unit>(), __ => action());
                    }

                    SetNeedsRegeneratingAndPrioritized(() => @out.Add(1));
                    SetNeedsRegeneratingAndPrioritized(() => SetNeedsRegeneratingAndPrioritized(() => @out.Add(4)));
                    SetNeedsRegeneratingAndPrioritized(() => @out.Add(2));
                    SetNeedsRegeneratingAndPrioritized(
                        () => SetNeedsRegeneratingAndPrioritized(
                            () => SetNeedsRegeneratingAndPrioritized(() => @out.Add(6))));
                    SetNeedsRegeneratingAndPrioritized(() => SetNeedsRegeneratingAndPrioritized(() => @out.Add(5)));
                    trans.Prioritized(new Node<Unit>(), trans2 => @out.Add(3));

                    return Unit.Value;
                },
                false);
            CollectionAssert.AreEqual(new[] { 1, 2, 3, 4, 5, 6 }, @out);
        }
    }
}

File: dotnet/src/Sodium/FSharp/Sodium.FSharp.Frp.Tests.Internal/CellTests.fs
module Sodium.Frp.Tests.Internal.Cell

open System.Collections.Generic
open NUnit.Framework
open Sodium.Frp

[<TestFixture>]
type ``Cell Tests``() =

    [<Test>]
    member __.``Test Transaction``() =
        let mutable calledBack = false
        TransactionInternal.Apply
            (
                (fun trans _ ->
                    trans.Prioritized (Node<unit>.Null, fun trans -> calledBack <- true)),
                false
            )
        Assert.IsTrue calledBack
    
    [<Test>]
    member __.``Test Regen``() =
        let out = List<_>()
        TransactionInternal.Apply
            (
                (fun trans _ ->
                    let setNeedsRegeneratingAndPrioritized a =
                        trans.SetNeedsRegenerating ()
                        trans.Prioritized (Node<unit> (), (fun _ -> a ()))
                    setNeedsRegeneratingAndPrioritized (fun () -> out.Add 1)
                    setNeedsRegeneratingAndPrioritized (fun () ->
                        setNeedsRegeneratingAndPrioritized (fun () -> out.Add 4))
                    setNeedsRegeneratingAndPrioritized (fun () -> out.Add 2)
                    setNeedsRegeneratingAndPrioritized (fun () ->
                        setNeedsRegeneratingAndPrioritized (fun () ->
                            setNeedsRegeneratingAndPrioritized (fun () -> out.Add 6)))
                    setNeedsRegeneratingAndPrioritized (fun () ->
                        setNeedsRegeneratingAndPrioritized (fun () -> out.Add 5))
                    trans.Prioritized (Node<unit> (), fun _ -> out.Add 3)),
                false
            )
        CollectionAssert.AreEqual ([1;2;3;4;5;6], out)